sgl-kernel/csrc/moe/fused_qknorm_rope_kernel.cu

```diff
@@ -195,43 +195,49 @@ __global__ void fusedQKNormRopeKernel(
   float cos_vals[numElemsPerThread];
   float sin_vals[numElemsPerThread];
   float pos_id = static_cast<float>(position_ids[tokenIdx]);
+  int const rotary_lanes = rotary_dim / numElemsPerThread;  // rotary range
+  bool const applyRotary = (laneId < rotary_lanes);
+  if (applyRotary) {
+    if constexpr (interleave) {
+      // Perform interleaving. Fill cos_vals and sin_vals.
+      for (int i = 0; i < numElemsPerThread; i++) {
+        elements2[i] = (i % 2 == 0) ? -elements[i + 1] : elements[i - 1];
+
+        int dim_idx = laneId * numElemsPerThread + i;
+        int half_dim = dim_idx / 2;
+        float freq = compute_freq_yarn(base, rotary_dim, half_dim, factor, low, high);
+        float theta = pos_id * freq;
+        __sincosf(theta, &sin_vals[i], &cos_vals[i]);
+      }
 
-  if constexpr (interleave) {
-    // Perform interleaving. Fill cos_vals and sin_vals.
-    for (int i = 0; i < numElemsPerThread; i++) {
-      if (i % 2 == 0) {
-        elements2[i] = -elements[i + 1];
-      } else {
-        elements2[i] = elements[i - 1];
+    } else {
+      // Neox style
+      // Before data exchange with in warp, we need to sync.
+      __syncwarp();
+      int const half_rotary_lanes = rotary_lanes / 2;
+      unsigned int active_mask = (1u << rotary_lanes) - 1;
+      for (int i = 0; i < numElemsPerThread; i++) {
+        elements2[i] = __shfl_xor_sync(active_mask, elements[i], half_rotary_lanes);
```

Comment:
Here needs to have the condition that half_rotary_lanes is power of 2. For example if rotary_lanes = 24 and half = 12, then lane ^ 12 will result in incorrect pairing (not just a simple front-back half swap), which will cause RoPE to fail. We have to handle the corner case.